src/utils.c

```diff
@@ -363,3 +308,53 @@ SEXP coerceUtf8IfNeeded(SEXP x) {
   return(ans);
 }
 
+static inline SEXP CLASSOF(SEXP x) {
+  switch(TYPEOF(x)) {
+  case NILSXP: {
+    return(ScalarString(char_NULL));
+  } break;
+  case LGLSXP: {
+    return(ScalarString(char_logical));
+  } break;
+  case INTSXP: {
+    return(ScalarString(INHERITS(x, char_ITime) ? char_ITime : (INHERITS(x, char_IDate) ? char_IDate : (INHERITS(x, char_ordered) ? char_ordered : (INHERITS(x, char_factor) ? char_factor : char_integer)))));
+  } break;
+  case REALSXP: {
+    return(ScalarString(Rinherits(x, char_nanotime) ? char_nanotime : (Rinherits(x, char_integer64) ? char_integer64 : (INHERITS(x, char_POSIXct) ? char_POSIXct : (INHERITS(x, char_Date) ? char_Date : char_double)))));
+  } break;
+  case STRSXP : {
+    return(ScalarString(char_character));
+  } break;
+  case CPLXSXP : {
+    return(ScalarString(char_complex));
+  } break;
+  case RAWSXP : {
+    return(ScalarString(char_raw));
+  } break;
+  case VECSXP : {
+    return(ScalarString(INHERITS(x, char_datatable) ? char_datatable : (INHERITS(x, char_dataframe) ? char_dataframe : char_list)));
+  } break;
+  default:
+    error(_("Type %s is not supported."), type2char(TYPEOF(x))); // # nocov
+  }
+  return(ScalarString(NA_STRING)); // # nocov
+}
+SEXP coerceAs(SEXP x, SEXP as, SEXP copyArg) {
+  // copyArg does not update in place, but only IF an object is of the same type-class as class to be coerced, it will return with no copy
+  if (!isVectorAtomic(x))
+    error("'x' is not atomic");
+  if (!isVectorAtomic(as))
+    error("'as' is not atomic");
+  if (isNewList(as))
+    error("'as' is a list");
+  int protecti=0;
+  bool copy = LOGICAL(copyArg)[0];
+  if (!copy && TYPEOF(x)==TYPEOF(as) && CLASSOF(x)==CLASSOF(as))
```

Comment:
doesn't `CLASSOF` equality imply `TYPOF` equality?

Reply:
yes but it is more expensive, this needs improvement anyway because any non handled class will be returned as just integer/double/etc. So two different (unhandled) classes will be interpretted as equal.

Reply:
OK agree